pkg/daemon/handler.go

```diff
@@ -0,0 +1,87 @@
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"go.uber.org/zap"
+	authv1 "k8s.io/api/authentication/v1"
+)
+
+const TokenHeaderKey = "Token"
+
+type TokenValidator interface {
+	Validate(ctx context.Context, token string) (*authv1.TokenReview, error)
+}
+
+type Freezer interface {
+	Freeze(ctx context.Context, podName, containerName string) error
+}
+
+type Thawer interface {
+	Thaw(ctx context.Context, podName, containerName string) error
+}
+
+type FreezeThawer interface {
+	Freezer
+	Thawer
+}
+
+type Handler struct {
+	Validator TokenValidator
+	Freezer   Freezer
+	Thawer    Thawer
+	Logger    *zap.SugaredLogger
+}
+
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	token := r.Header.Get(TokenHeaderKey)
+	if token == "" {
+		h.Logger.Error("No token in header")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	resp, err := h.Validator.Validate(r.Context(), token)
+	if err != nil {
+		h.Logger.Error("Validating token failed")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	if !resp.Status.Authenticated {
+		h.Logger.Error("Authenticating via token failed")
+		w.WriteHeader(http.StatusForbidden)
+		return
+	}
+
+	var m messageBody
+	decoder := json.NewDecoder(r.Body)
+	err = decoder.Decode(&m)
+	if err != nil {
+		http.Error(w, "error decoding", http.StatusBadRequest)
+	}
+
+	switch m.Action {
+	case "pause":
+		h.Logger.Info("pause request received, freezing pod...")
+		h.Freezer.Freeze(r.Context(), resp.Status.User.Extra["authentication.kubernetes.io/pod-uid"][0], "user-container")
```

Comment:
this assumes the default user container name which can be overridden, so we'll need to unhardcode this